### PR TITLE
Add service to generate payload for officeconnector oneoffixx action

### LINKED ISSUE #4137: OneOffixx: oc_oneoffixx Service
Similiar to OfficeConnectorCheckoutPayload.

Lives on the site root and operates via the document UUID.
https://github.com/4teamwork/opengever.core/blob/c723c47a8e668b9f9b35e9ba827a330e77260ba1/opengever/officeconnector/configure.zcml#L61-L68

https://github.com/4teamwork/opengever.core/blob/c723c47a8e668b9f9b35e9ba827a330e77260ba1/opengever/officeconnector/service.py#L88-L128

Includes `oneoffixx-connect-xml-url` defined in #4135

### PR DESCRIPTION
**Pull request by njohner**
Add service to generate payload for the officeconnector oneoffixx action. The payload contains the necessary information to retrieve the connect XML (`connect-xml` key)

resolves #4137

### CODE REVIEW

File: opengever/officeconnector/service.py [hunk 1]
```diff
@@ -228,3 +236,42 @@ def render(self):
...
 
         return json.dumps(payloads)
+
+class OfficeConnectorOneOffixxPayload(OfficeConnectorPayload):
+    """Issue JSON instruction payloads for OfficeConnector.
+
+    Consists of the minimal instruction set with which to perform a full
+    checkout checkin cycle for a file attached to a document.
+    """
+    @staticmethod
```
Comment (Rotonen) [inline opengever/officeconnector/service.py]: Rather module scope the method, make a mixin or accept the duplication.

File: opengever/officeconnector/service.py [hunk 2]
```diff
@@ -129,12 +133,9 @@ def get_base_payloads(self):
             for document in documents:
                 payloads.append(
                     {
-                        'content-type': document.get_file().contentType,
```
Comment (Rotonen) [inline opengever/officeconnector/service.py]: Why do these get moved?

File: opengever/officeconnector/service.py [hunk 3]
```diff
@@ -228,3 +236,42 @@ def render(self):
...
+        for payload in payloads:
+            # A permission check to verify the user is also able to upload
+            authorized = api.user.has_permission(
+                'Modify portal content',
+                obj=payload['document'],
+                )
+
+            if authorized:
+                del payload['document']
+                payload['checkin-with-comment'] = '@@checkin_document'
```
Comment (Rotonen) [inline opengever/officeconnector/service.py]: We do not need checkin or upload data in the oneoffixx payload. We'll need information on where the XML is, how to unshadow and then the checkout URL (JWT) of the document to skip a request and directly fetch the normal checkout payload.

File: opengever/officeconnector/service.py [hunk 4]
```diff
@@ -108,6 +108,10 @@ def __init__(self, context, request):
         super(OfficeConnectorPayload, self).__init__(context, request)
         self.uuids = json.loads(request['BODY'])
 
+    @staticmethod
```
[the remark this answers is not in the document]
Comment (njohner) [reply opengever/officeconnector/service.py]: I did that to overwrite it in `OfficeConnectorOneOffixxPayload` which uses this one as a baseclass, but has a different condition for documents being valid. 
So module scope would not work here. Duplication would mean duplicate the whole baseclass

**Comment by njohner** [comment]
Ok can you have another look?
The payload now contains:
* where to find the XML
* the checkout url JWT
* the filename
* the document url
I'm not passing anything for the unshadowing. I think this should not be necessary, as the checkin will unlock the document and the document unlock also removes the shadow state. (https://github.com/4teamwork/opengever.core/blob/master/opengever/document/document.py#L334)

File: opengever/officeconnector/service.py [hunk 5]
```diff
@@ -230,3 +256,43 @@ def render(self):
         self.request.response.setHeader('Content-type', 'application/json')
 
         return json.dumps(payloads)
+
+
+class OfficeConnectorOneOffixxPayload(OfficeConnectorPayload):
+    """Issue JSON instruction payloads for OfficeConnector.
```
Comment (Rotonen) [inline opengever/officeconnector/service.py]: I think we need a fresh docstring here.

File: opengever/officeconnector/service.py [hunk 6]
```diff
@@ -230,3 +256,43 @@ def render(self):
...
+    """
+    @staticmethod
+    def document_is_valid(document):
+        return document and document.is_shadow_document()
+
+    def render(self):
+        payloads = self.get_base_payloads()
+
+        # Upload API will be included as a registry setting in the future when
+        # the plone.api endpoint gets made - for now we've made a custom upload
```
Comment (Rotonen) [inline opengever/officeconnector/service.py]: This is not correct at the moment. We're pending TUS support on the client side, `plone.restapi` already has support baked in.

http://plonerestapi.readthedocs.io/en/latest/tusupload.html

Sidenote: `plone.api` is also the wrong product. I know we named it badly, but people objected to having called it `plone.ipa` in 2011, even though it's about the inverse of what an api is (maybe that had something to do with it having been birthed in the belgian beer sprint, see https://www.coactivate.org/projects/belgian-beer-sprint/project-home for some of the the history). It's essentially just a collection of shorthand utils for most commonly done things deemed too Zope-y and repetitive.

File: opengever/officeconnector/service.py [hunk 7]
```diff
@@ -230,3 +256,43 @@ def render(self):
...
+            # A permission check to verify the user is also able to upload
+            authorized = api.user.has_permission(
+                'Modify portal content',
+                obj=payload['document'],
+                )
+
+            if authorized:
+                document = payload['document']
+                checkout_token = create_oc_url(self.request, document, {"action": "checkout"})
+                payload['filename'] = IAnnotations(document).get("filename")
```
Comment (Rotonen) [inline opengever/officeconnector/service.py]: We also need to do this conditionally in the normal checkout payload for documents if they're in the shadow state so the OC-side logic of avoiding to download when we already have a file kicks in.

File: opengever/officeconnector/service.py (hunk 6 above)
Comment (njohner) [reply opengever/officeconnector/service.py]: That was a copy-paste fail from the Checkout Payload. That comment should not be here anyway as we do not include the upload information in the OneOffixx payload. But hen maybe that comment should be corrected in the Checkout payload...

File: opengever/oneoffixx/browser/form.py [hunk 8]
```diff
@@ -153,17 +153,9 @@ def finish_document_creation(self, data):
...
+        """Add the oneoffixx external_editor URL to redirector queue.
         """
-        # Check out the new document
-        manager = self.context.restrictedTraverse('checkout_documents')
-
-        # When it will be implemented we should use the new oneoffix
-        # action instead of checkout
-        # we can extend setup_external_edit_redirect to take a keyword argument
-        # action
-        manager.checkout(new_doc)
```
Comment (njohner) [inline opengever/oneoffixx/browser/form.py]: The checkout will be done from OfficeConnector

File: opengever/officeconnector/helpers.py [hunk 9]
```diff
@@ -34,25 +34,25 @@ def parse_bcc(request):
 
 def parse_documents(request, context, action):
     documents = []
-
     if (
             request['REQUEST_METHOD'] == 'GET'
             or request['REQUEST_METHOD'] == 'POST'
-            and 'BODY' not in request
+            and ('BODY' not in request
+                 or isinstance(json.loads(request['BODY']), list))
```
Comment (njohner) [inline opengever/officeconnector/helpers.py]: When fetching the checkout url from the OneOffixxPayload, the request is POST but has a BODY, but it's still the context that is the document... In that case the BODY is a list and not a dict in the `elif` below. If you have a cleaner solution here...

Comment (Rotonen) [reply opengever/officeconnector/helpers.py]: Not on this schedule. That'll do for now.

**Pull request merged**